File: hospital-ai/backend/app/schemas/patient.py
"""Patient domain schemas."""

from datetime import date, datetime
from enum import Enum
from typing import Any, List, Optional
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field, field_validator


class PatientGender(str, Enum):
    MALE = "Male"
    FEMALE = "Female"
    OTHER = "Other"


class BloodGroup(str, Enum):
    A_POS = "A+"
    A_NEG = "A-"
    B_POS = "B+"
    B_NEG = "B-"
    AB_POS = "AB+"
    AB_NEG = "AB-"
    O_POS = "O+"
    O_NEG = "O-"


class PatientBase(BaseModel):
    first_name: str = Field(min_length=1, max_length=100)
    last_name: str = Field(min_length=1, max_length=100)
    date_of_birth: date
    gender: PatientGender
    blood_group: Optional[BloodGroup] = None
    phone: str = Field(min_length=7, max_length=20)
    email: Optional[EmailStr] = None
    address: Optional[str] = Field(default=None, max_length=500)
    city: Optional[str] = Field(default=None, max_length=100)
    state: Optional[str] = Field(default=None, max_length=100)
    country: Optional[str] = Field(default="India", max_length=100)
    emergency_contact_name: Optional[str] = Field(default=None, max_length=150)
    emergency_contact_phone: Optional[str] = Field(default=None, max_length=20)
    allergies: Optional[str] = Field(default=None, max_length=2000)
    medical_history: Optional[str] = Field(default=None, max_length=5000)
    current_medications: Optional[str] = Field(default=None, max_length=2000)
    insurance_provider: Optional[str] = Field(default=None, max_length=150)
    insurance_number: Optional[str] = Field(default=None, max_length=100)

    @field_validator("first_name", "last_name", "phone")
    @classmethod
    def strip_required(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field cannot be empty")
        return cleaned

    @field_validator("date_of_birth")
    @classmethod
    def validate_dob(cls, value: date) -> date:
        if value > date.today():
            raise ValueError("Date of birth cannot be in the future")
        if value.year < 1900:
            raise ValueError("Date of birth is unrealistically old")
        return value


class PatientCreate(PatientBase):
    pass


class PatientUpdate(BaseModel):
    first_name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    last_name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    date_of_birth: Optional[date] = None
    gender: Optional[PatientGender] = None
    blood_group: Optional[BloodGroup] = None
    phone: Optional[str] = Field(default=None, min_length=7, max_length=20)
    email: Optional[EmailStr] = None
    address: Optional[str] = Field(default=None, max_length=500)
    city: Optional[str] = Field(default=None, max_length=100)
    state: Optional[str] = Field(default=None, max_length=100)
    country: Optional[str] = Field(default=None, max_length=100)
    emergency_contact_name: Optional[str] = Field(default=None, max_length=150)
    emergency_contact_phone: Optional[str] = Field(default=None, max_length=20)
    allergies: Optional[str] = Field(default=None, max_length=2000)
    medical_history: Optional[str] = Field(default=None, max_length=5000)
    current_medications: Optional[str] = Field(default=None, max_length=2000)
    insurance_provider: Optional[str] = Field(default=None, max_length=150)
    insurance_number: Optional[str] = Field(default=None, max_length=100)

    @field_validator("date_of_birth")
    @classmethod
    def validate_dob(cls, value: Optional[date]) -> Optional[date]:
        if value is None:
            return value
        if value > date.today():
            raise ValueError("Date of birth cannot be in the future")
        if value.year < 1900:
            raise ValueError("Date of birth is unrealistically old")
        return value


class PatientAIExtensions(BaseModel):
    """
    Extension points for future AI agents.
    Always present on patient payloads; currently unused (null).
    """

    ai_context: Optional[Any] = None
    latest_diagnosis: Optional[str] = None
    latest_report: Optional[Any] = None
    prediction_history: Optional[Any] = None


class PatientResponse(PatientBase, PatientAIExtensions):
    id: UUID
    patient_number: str
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PatientListResponse(BaseModel):
    items: List[PatientResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class MessageResponse(BaseModel):
    message: str

File: hospital-ai/frontend/src/types/patient.ts
export type PatientGender = 'Male' | 'Female' | 'Other';

export type BloodGroup =
  | 'A+'
  | 'A-'
  | 'B+'
  | 'B-'
  | 'AB+'
  | 'AB-'
  | 'O+'
  | 'O-';

/** Reserved fields for future AI agents — currently unused. */
export interface PatientAIExtensions {
  ai_context: unknown | null;
  latest_diagnosis: string | null;
  latest_report: unknown | null;
  prediction_history: unknown | null;
}

export interface Patient extends PatientAIExtensions {
  id: string;
  patient_number: string;
  first_name: string;
  last_name: string;
  date_of_birth: string;
  gender: PatientGender;
  blood_group: BloodGroup | null;
  phone: string;
  email: string | null;
  address: string | null;
  city: string | null;
  state: string | null;
  country: string | null;
  emergency_contact_name: string | null;
  emergency_contact_phone: string | null;
  allergies: string | null;
  medical_history: string | null;
  current_medications: string | null;
  insurance_provider: string | null;
  insurance_number: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface PatientFormValues {
  first_name: string;
  last_name: string;
  date_of_birth: string;
  gender: PatientGender | '';
  blood_group: BloodGroup | '';
  phone: string;
  email: string;
  address: string;
  city: string;
  state: string;
  country: string;
  emergency_contact_name: string;
  emergency_contact_phone: string;
  allergies: string;
  medical_history: string;
  current_medications: string;
  insurance_provider: string;
  insurance_number: string;
}

export interface PatientListParams {
  page?: number;
  page_size?: number;
  search?: string;
  gender?: PatientGender | '';
  blood_group?: BloodGroup | '';
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface PatientListResponse {
  items: Patient[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export const GENDER_OPTIONS: PatientGender[] = ['Male', 'Female', 'Other'];

export const BLOOD_GROUP_OPTIONS: BloodGroup[] = [
  'A+',
  'A-',
  'B+',
  'B-',
  'AB+',
  'AB-',
  'O+',
  'O-',
];

export const EMPTY_PATIENT_FORM: PatientFormValues = {
  first_name: '',
  last_name: '',
  date_of_birth: '',
  gender: '',
  blood_group: '',
  phone: '',
  email: '',
  address: '',
  city: '',
  state: '',
  country: 'India',
  emergency_contact_name: '',
  emergency_contact_phone: '',
  allergies: '',
  medical_history: '',
  current_medications: '',
  insurance_provider: '',
  insurance_number: '',
};
